test/p_test.c

```diff
@@ -109,6 +115,45 @@ static int p_get_params(void *provctx, OSSL_PARAM params[])
                 strcpy(p->data, buf);
             else
                 ok = 0;
+        } else if (strcmp(p->key, "digest-check") == 0) {
+            unsigned int digestsuccess = 0;
+
+            /*
+             * Test we can use an algorithm from another provider. We're using
+             * legacy to check that legacy is actually available and we haven't
+             * just fallen back to default.
+             */
+#ifdef PROVIDER_INIT_FUNCTION_NAME
+            EVP_MD *md4 = EVP_MD_fetch(ctx->libctx, "MD4", NULL);
+            EVP_MD_CTX *mdctx = EVP_MD_CTX_new();
+            const char *msg = "Hello world";
+            unsigned char out[16];
+
+            /*
+             * We should have the default provider available that we loaded
+             * ourselves, and the legacy provider which we inherit from the
+             * parent libctx. We should also have "this" provider available.
+             */
+            if (OSSL_PROVIDER_available(ctx->libctx, "default")
```

Comment:
Is diagnostic output possible here to distinguish the failure cases?

Reply:
I'm not sure how to emit diagnostic output from inside this test provider?

Reply:
What about raising an error?

Reply:
I've raised different error codes for different types of failure cases.